dspace-api/src/main/java/org/dspace/discovery/SolrServiceImpl.java

```diff
@@ -860,141 +857,154 @@ protected SolrQuery resolveToSolrQuery(Context context, DiscoverQuery discoveryQ
         return solrQuery;
     }
 
-    protected DiscoverResult retrieveResult(Context context, DiscoverQuery query, QueryResponse solrQueryResponse)
-        throws SQLException {
-        DiscoverResult result = new DiscoverResult();
-
-        if (solrQueryResponse != null) {
-            result.setSearchTime(solrQueryResponse.getQTime());
-            result.setStart(query.getStart());
-            result.setMaxResults(query.getMaxResults());
-            result.setTotalSearchResults(solrQueryResponse.getResults().getNumFound());
-
-            List<String> searchFields = query.getSearchFields();
-            for (SolrDocument doc : solrQueryResponse.getResults()) {
-                IndexableObject indexableObject = findIndexableObject(context, doc);
-
-                if (indexableObject != null) {
-                    result.addIndexableObject(indexableObject);
-                } else {
-                    log.error(LogManager.getHeader(context,
-                            "Error while retrieving DSpace object from discovery index",
-                           "Unique identifier: " + doc.getFirstValue(SearchUtils.RESOURCE_UNIQUE_ID)));
-                    continue;
-                }
-
-                DiscoverResult.SearchDocument resultDoc = new DiscoverResult.SearchDocument();
-                //Add information about our search fields
-                for (String field : searchFields) {
-                    List<String> valuesAsString = new ArrayList<>();
-                    for (Object o : doc.getFieldValues(field)) {
-                        valuesAsString.add(String.valueOf(o));
+    protected DiscoverResult retrieveResult(Context context, DiscoverQuery query)
+        throws SQLException, SolrServerException, IOException, SearchServiceException {
+        boolean valid = false;
+        DiscoverResult result = null;
+        SolrQuery solrQuery = resolveToSolrQuery(context, query);
+        while (!valid) {
```

Comment:
There's no limit to how often this loop occurs. Depending on how likely this problem occurs, this loop may keep on being iterated for a long time until you get a search where all incorrect documents have been removed

Reply:
I agree with @benbosman on this.  I think there should be some sort of limit to the number of times this loop is run.  We could limit it to running 10 times max. If it bypasses that limit, we throw (and log) an error that says something like: "The Discovery (Solr) index has a large number of stale entries, and we could not complete this request. Please reindex all content to remove these stale entries (e.g. dspace index-discovery -f)."   

I would hope this limit would never be hit (as this seems like something that should occur rarely), but it'd ensure there is a "timeout" to this request.